jwst/background/background_sub_wfss.py

```diff
@@ -0,0 +1,283 @@
+            Initial mask to be applied to the data, True where bad.
+            Typically this would mask out the real sources in the data.
+
+        Returns
+        -------
+        float
+            Scaling factor that minimizes sci - factor*bkg,
+            taking into account residuals and outliers.
+        ndarray[bool]
+            Outlier mask generated by the iterative clipping procedure.
+        """
+        if mask is None:
+            mask = np.zeros(sci.shape, dtype="bool")
+        self._update_nans(sci, bkg, var, mask)
+
+        # iteratively reject more and more outliers
+        i = 0
+        last_rms_resid = np.inf
+        while (i < self.maxiter):
+
+            # compute the factor that minimizes the residuals
+            factor = self.err_weighted_mean(sci, bkg, var)
+            sci_sub = sci-factor*bkg
+
+            # Check fractional improvement stopping criterion before incrementing.
+            # Note this never passes in iteration 0 because last_rms_resid is inf.
+            if self.delta_rms_thresh > 0:
+                rms_resid = self._compute_rms_residual(sci_sub)
+                with warnings.catch_warnings():
+                    warnings.filterwarnings("ignore",
+                                            category=RuntimeWarning,
+                                            message="invalid value encountered in scalar divide")
+                    fractional_diff = (last_rms_resid - rms_resid)/last_rms_resid
+                if fractional_diff < self.delta_rms_thresh:
+                    msg = (f"Stopping at iteration {i}; too little improvement "
+                           "since last iteration (hit delta_rms_thresh).")
+                    log.info(msg)
+                    break
+                last_rms_resid = rms_resid
+
+            i += 1
+
+            # Reject outliers based on residual between sci and bkg.
+            # Updating the sci, var, and bkg nan values means that
+            # they are ignored by nanpercentile in the next iteration
+            limits = np.nanpercentile(sci_sub, (self.p, 100-self.p))
+            mask += np.logical_or(sci_sub < limits[0], sci_sub > limits[1])
+            self._update_nans(sci, bkg, var, mask)
+
+        if i >= self.maxiter:
+            log.info(f"Stopped at maxiter ({i}).")
+
+        self._iters_run_last_call = i
+        return self.err_weighted_mean(sci, bkg, var), mask
+
+
+    def err_weighted_mean(self, sci, bkg, var):
+        """Remove any var=0 values, which can happen for real data"""
+        mask = (var == 0)
+        self._update_nans(sci, bkg, var, mask)
+        return np.nansum(sci*bkg/var) / np.nansum(bkg*bkg/var)
+
+
+    def _update_nans(self, sci, bkg, var, mask):
+        sci[mask] = np.nan
+        bkg[mask] = np.nan
+        var[mask] = np.nan
+
+
+    def _compute_rms_residual(self, sci_sub):
+        """
+        Calculate the RMS of the cross-dispersion background, which is found by taking
```

Comment:
Calculate the background-subtracted RMS along the dispersion axis, which is found by taking the median profile of the image collapsed along the cross-dispersion axis.